### ISSUE TITLE
401 errors when config is cached

### ISSUE THREAD

**Issue by marnickmenting** [userlist/userlist-laravel#1]
I haven't put my finger on this exactly, but everytime I run `php artisan config:cache` I get 401 Unauthorized errors when running `php artisan userlist:import`. When I run `php artisan config:clear` the error is gone.

This is the error:
```
Client error: `POST https://push.userlist.com/companies` resulted in a `401 Unauthorized` response: {"status":401,"code":"unauthorized","errors":["You are not authorized to do this."]}
```

I had similar problems in the past with using env() in the code, but I don't see any wrong uses in the source.

I have USERLIST_PUSH_ID and USERLIST_PUSH_KEY in .env file and my config.php is (yes, a company and user is the same model):
```
<?php

return [
    'user_model' => \App\Models\Website::class,
    'company_model' => \App\Models\Website::class,
    'push_id' => env('USERLIST_PUSH_ID'),
    'push_key' => env('USERLIST_PUSH_KEY')
];
```

laravel/framework: 8.27.0
userlist/userlist-laravel: dev-master

**Comment by benedikt** [comment]
Hey @marnickmenting,

thanks for reporting this. How are you creating instances of `\Userlist\Push`? Are you creating new instances or using the singleton provided by userlist-laravel via the dependency injection mechanism?

If you're manually creating new instances using `$userlist = new \Userlist\Push();`, please be sure to pass the Userlist config like this: `$userlist = new \Userlist\Push(config('userlist'));`

Does that help and fix the problem?

Cheers,

Benedikt

**Comment by marnickmenting** [comment]
I am not using the Push instance at the moment, only syncing using userlist:import command.

I have a class extending from CompanyTransform and one from UserTransform that both have a public transform function:

```
<?php

namespace App\Services\Userlist;

use Userlist\Laravel\Services\UserTransform as Transform;

class UserTransform extends Transform {
    public function transform($website) {
        // method to switch to tenant database connection mode
        change_tenant($website);

        return [
            // Bunch of properties
        ];
    }
}

```

They are loaded in AppServiceProvider::boot by this code:

```
$this->app->bind('Userlist\Laravel\Contracts\UserTransform', 'App\Services\Userlist\UserTransform');
$this->app->bind('Userlist\Laravel\Contracts\CompanyTransform', 'App\Services\Userlist\CompanyTransform');
```

**Comment by benedikt** [comment]
I tried to reproduce the issue, but wasn't able to do so. Can you double check that your `bootstrap/cache/config.php` includes the Userlist credentials after running `php artisan config:cache`?

**Comment by marnickmenting** [comment]
I did some debugging, I found that in UserlistServiceProvider.php on line 46 the config('userlist') return null. Even the $app variable returns null. But this is the case with or without cache, which is weird.

```
/**
     * Register any application services.
     *
     * @return void
     */
    public function register()
    {
        $this->app->singleton('\Userlist\Push', function($app) {
            Log::debug(print_r($app,true )); // result: null
            Log::debug(print_r(config('userlist'), true)); // result: null
            return new \Userlist\Push(config('userlist'));
        });
    }
```

**Comment by marnickmenting** [comment]
> I tried to reproduce the issue, but wasn't able to do so. Can you double check that your `bootstrap/cache/config.php` includes the Userlist credentials after running `php artisan config:cache`?

The credentials are indeed present in cache/config.php.

**Comment by benedikt** [comment]
Where do you store your Userlist configuration? You mention `config.php`, but it should be `config/userlist.php` for the namespace lookup to work.

**Comment by marnickmenting** [comment]
Yes, the config is in config/userlist.php, but the cached version is in bootstrap/cache/config.php

**Comment by benedikt** [comment]
I'm a bit at a loss right now. I don't see a reason why `config('userlist')` would not return the correct value.

**Comment by benedikt** [comment]
Closed by 9d5d4b0a7e6557cd402fb34691a36bcadead993f
